src/main/java/co/cask/spanner/source/SpannerSource.java

```diff
@@ -0,0 +1,283 @@
+    try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+         ObjectOutputStream objectOutputStream = new ObjectOutputStream(byteArrayOutputStream)) {
+      objectOutputStream.writeObject(object);
+      objectOutputStream.flush();
+      String objectBytesString = Base64.getEncoder().encodeToString(byteArrayOutputStream.toByteArray());
+      configuration.set(propertyName, objectBytesString);
+    }
+  }
+
+  private PartitionOptions getPartitionOptions() {
+    PartitionOptions.Builder builder = PartitionOptions.newBuilder();
+    if (config.partitionSizeBytes != null) {
+      builder.setPartitionSizeBytes(config.partitionSizeBytes);
+    }
+    if (config.maxPartitions != null) {
+      builder.setMaxPartitions(config.maxPartitions);
+    }
+    return builder.build();
+  }
+
+  /**
+   * Endpoint method to get the output schema of spanner query.
+   *
+   * @param request {@link GetSchemaRequest} containing information required for connection and query to execute.
+   * @return schema of fields
+   */
+  @Path("getSchema")
+  public Schema getSchema(GetSchemaRequest request) throws Exception {
+    Spanner spanner = SpannerUtil.getSpannerService(request.serviceFilePath, request.project);
+    DatabaseClient databaseClient =
+      spanner.getDatabaseClient(DatabaseId.of(request.project, request.instance, request.database));
+    ResultSet resultSet = databaseClient.singleUse().executeQuery(Statement.of(request.query));
+    List<Schema.Field> schemaFields = new ArrayList<>();
+    if (resultSet.next()) {
+      List<Type.StructField> structFields = resultSet.getType().getStructFields();
+      for (Type.StructField field : structFields) {
+        // todo find out if column type is nullable from information schema.
+        // we assume all of them accept null values currently as
+        // StructField only provides type and name and doesn't have information if its optional
+        schemaFields.add(Schema.Field.of(field.getName(),
+                                         Schema.nullableOf(getSchemaForCode(field.getType().getCode()))));
+      }
+    }
+    return Schema.recordOf("outputSchema", schemaFields);
+  }
+
+  /**
+   * get {@link Schema} for the {@link Type.Code},
+   */
+  private Schema getSchemaForCode(Type.Code code) throws SQLException {
+    switch (code) {
+      case BOOL:
+        return Schema.of(Schema.Type.BOOLEAN);
+      case INT64:
+        return Schema.of(Schema.Type.LONG);
+      case FLOAT64:
+        return Schema.of(Schema.Type.DOUBLE);
+      case STRING:
+        return Schema.of(Schema.Type.STRING);
+      case BYTES:
+        return Schema.of(Schema.Type.BYTES);
+      // todo support the following types
+      case ARRAY:
+      case DATE:
+      case TIMESTAMP:
+      case STRUCT:
+      default:
+        throw new SQLException(new UnsupportedTypeException("Unsupported SQL Type: " + code.name()));
+    }
+  }
+
+  class GetSchemaRequest {
```

Comment:
you can just use the source config as the request object. If you have this separate object, there is a hidden requirement that it has the exact same field names.